### PR TITLE
Add unique constraint on database [name + engine]

### LINKED ISSUE #11106: We should add unique constraint on Database name + engine (driver)
**Describe the bug**
POST to /api/database will overwrite an existing database instead of erroring

**To Reproduce**
Create a database with a POST to /api/database.
Run the same call again.

**Expected behavior**
Expect Metabase to error and say, "Database with name 'foo' already exists"

**Actual behavior**
Metabase gives another 200 and will overwrite any previous database config with the same name.

**Information about your Metabase Installation:**
Metabase 0.33.3

### PR DESCRIPTION
**Pull request by mohammedamarnah**
Fixes #11106

The issue was happening because there weren't any constraints/checks on the uniqueness of database names across engines. This PR adds a test function that checks if a pair [db-name engine] exists or not, and returns 400 with a message if it already exists.

Should we consider checking for name uniqueness on updating current databases as well?

##### PS: This is my first _serious_ contribution in any open-source project, so any comment would be valuable!

### CODE REVIEW

**Review by camsaul — CHANGES_REQUESTED** [review]
*  We should make this a DB-level constraint. See [this migration](https://github.com/metabase/metabase/blob/master/resources/migrations/000_migrations.yaml#L5208) for an example where we did something similar

**Comment by kdoh** [comment]
Hi @mohammedamarnah. Congrats on your first OSS contribution! That's very exciting. Sorry it took us a while to get to reviewing it.

Since it has been a while since the initial PR was opened I just wanted to follow up and confirm that you'd still be willing to make the changes @camsaul suggested. Let us know either way or if you need more information.

**Comment by mohammedamarnah** [comment]
Hello, @kdoh! Thank you so much :)

I would love to make the changes!
I was just trying to figure out some more information about running the migrations. I come from a rails background, so running DB migrations there is pretty simple. Do you have a guide on that?
Thank you again! :)

**Comment by mohammedamarnah** [comment]
@camsaul Hello! :)
I added the migration and I think running `lein test` ran the migrations and I think its working, but I'm not sure if this is right or not. :sweat_smile:

**Comment by flyingmachine** [comment]
@mohammedamarnah this is looking great! Congrats again on your first contribution!

For the `constraintName` I think that instead of `idx_uniq_table_name_engine` we want to use `idx_uniq_database_name_engine`. Could you make that change?

**Comment by mohammedamarnah** [comment]
@flyingmachine Oh thank you so much! 
Of course! At first, I thought that `name` and `engine` were columns in `metabase_table`, and wrote everything based on that, but then realized that it's in `metabase_database` instead and changed everything accordingly but seems like I missed this one. 😅

**Comment by camsaul** [comment]
I think this is a good idea, and the changes look good to me at a glance, but we need to do some work updating our tests. A few of our tests make copies of the current test database so they can change the synced metadata without affecting other tests. I think most of those funnel thru our `with-temp-copy-of-db` macro so it would be straightforward to have it generate a unique name for the copies. (I think it would be easiest for a member of the Metabase team to tackle that work.)

Also for the next commit we need to add `[ci mysql] [ci postgres]` to the commit message so it runs against Postgres and MySQL.

**Review by camsaul — CHANGES_REQUESTED** [review]
LGTM but we need to fix tests (I or another member of the MB team should tackle this since it involves reworking some of our test scaffolding)

**Comment by mohammedamarnah** [comment]
@camsaul That's wonderful! Would you like me to edit anything else then?

**Comment by walterl** [comment]
@camsaul Is [this](https://github.com/metabase/metabase/pull/11228/files#diff-dd1ef3aafdf7b7ce7b51d32205694027) what you had in mind for the test scaffolding changes?

**Comment by camsaul** [comment]
This PR does look good overall but the one problem we have is when someone uses our tool to migrate from H2 -> MySQL/MariaDB/Postgres the new DB will have these constraints in place and if the old DB had some data that didn't follow those constraints the migration will fail (see related issue #10961, which arose after we added a constraint for to make user emails unique.)

I still think we should merge this PR but since it could cause issues with the migration tool I want to think of a way to resolve that issue as well

**Comment by camsaul** [comment]
Let me think about it for a bit and hopefully I can have a fix ready for the next release so we can fix #10961 and merge this without causing new issues

**Review by camsaul — CHANGES_REQUESTED** [review]
~~I want to address #10961 first before merging. I'm going to try to come up with fix this week and then I'll merge this afterwards~~

**Review by camsaul — CHANGES_REQUESTED** [review]
It actually looks like the syntax for the `preCondtions` is wrong here: see #11847 (we were using the wrong syntax in the migrations this was modeled on).

Before we can merge this, the syntax needs to be fixed and a test should get added to this PR to make sure the constraint actually works as expected. Add `[ci mysql]` and `[ci postgres]` to the commit message so tests will run against MySQL and Postgres. 

Once that's done, I can figure out how to get our H2->MySQL/Postgres tool working correctly in combination with this PR (address #10961 first before merging)

**Comment by mohammedamarnah** [comment]
@camsaul I think I added the test but I'm not sure if I did it the right way. Let me know if it should be done in some other way 😅.

File: test/metabase/models/database_test.clj [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+            [toucan.db :as db]))
+
+(defn- create-database!
+  [db-name engine]
+  (db/insert! Database
+        :name db-name
+        :engine (name engine)
+        :details {:db (join "-" (mapv lower-case (split db-name #"(?=[A-Z])")))}))
+
+(deftest disallow-duplicate-databases-with-same-name-and-engine
```
Comment (camsaul) [inline test/metabase/models/database_test.clj]: we should only run these if we're currently testing against that database, for example we want to be able to run tests against just `:h2` without having a MySQL or Postgres database running. What we usually do is 

```clj
(require '[metabase.driver :as driver] '[metabase.test :as mt])

(deftest my-test
  ;; test-drivers will run the body against :mysql, :postgres, etc. if tests are enabled for that DB
  (mt/test-drivers #{:mysql :postgres :h2}
    ;; driver/*driver* will be bound to :mysql, :postgres, etc.
    (testing driver/*driver*
      ...)))
```

**Review by camsaul — CHANGES_REQUESTED** [review]
need to tweak the tests so they only run against DBs when tests for those DBs are enabled

File: test/metabase/models/database_test.clj (hunk 1 above)
Comment (camsaul) [reply test/metabase/models/database_test.clj]: actually on second thought you don't really to test this against the different DB types, you can just write the test once, we run tests with all of our supported backing application database types anyway. But you should clear out the DB when you're done. Maybe a test like

```clj
(deftest my-test
  (try
    (create-database! "My DB")
    (is (thrown? ...))
    (finally
      (delete-database "My DB"))))
```

that should probably do it

**Comment by mohammedamarnah** [comment]
Hi everyone!
Any updates on this?

**Review by camsaul — APPROVED** [review]

**Comment by camsaul** [comment]
I fixed the merge conflicts and opened PR #13878 to rebase this on latest `master`, so assuming the tests pass we can go ahead and merge it

**Comment by camsaul** [comment]
Merged in as part of #13878

**Pull request closed without merging**